Tests/FunctionalTests/MixedTargetTests.swift

```diff
@@ -476,4 +479,28 @@ final class MixedTargetTests: XCTestCase {
             )
         }
     }
+
+    #else
+
+    // MARK: - Test Mixed Targets unsupported on non-macOS
+
+    func testMixedTargetOnlySupportedOnMacOS() throws {
+        try fixture(name: "MixedTargets/BasicMixedTargets") { fixturePath in
+            let commandExecutionError = try XCTUnwrap(
+                XCTAssertBuildFails(
+                    fixturePath,
+                    extraArgs: ["--target", "BasicMixedTarget"]
+                )
+            )
+
+            XCTAssert(
+                commandExecutionError.stderr.contains(
+                    "error: Targets with mixed language sources are only supported on Apple platforms."
```

Comment:
Does that mean that Swift/C++ targets are disabled on non-Darwin platforms too? What's the reason for that?

Reply:
Maybe the error message should be updated to not give that impression?
```suggestion
                    "error: Targets with mixed Swift/Objective-C language sources are only supported on Apple platforms."
```

Reply:
To be clear, this unit test is out of date as the impl. does not throw an error like this anymore (https://github.com/apple/swift-package-manager/pull/5919/commits/3202a825c8afcd527e3d808c3be5c2a3c047f263). But, I'm wondering if you think I should be throwing an error like this when the target contains Objective-C.

Reply:
IMO throwing an error makes sense. cc @neonichu

Reply:
Hm, I don't think we are throwing an error today? The thinking may have been to relegate that to the compiler instead of hardcoding in SwiftPM what the compiler does or doesn't support.

Reply:
I'm not sure if the compiler produces a sensible error for this though. I'd expect it'd be something about missing Objective-C or anything else of that nature that would confuse the users. @ncooke3 I think it's worth verifying what the current behavior is.

Reply:
When the Objective-C/Swift targets failed to build on Linux during the last CI run, the errors were that it couldn't resolve `#import <Foundation/Foundation.h>` and ObjC-interop related things like `@objc`. See these logs: https://ci.swift.org/job/swift-package-manager-self-hosted-Linux-smoke-test/2668/console

Reply:
Right, IMO those errors aren't user-friendly. SwiftPM preemptively throwing its own more readable error would save time for our users in their investigation, and they wouldn't have to wait for the late stage where a compiler is already is spawned to discover that this configuration is not supported.

Reply:
@MaxDesiatov, I agree. Though I realized that throwing an error would make it the following case impossible: a mixed lang. target containing Objc where the Objc files are guarded on Objc availability. In practice, such a setup might exist for a library that offers a public ObjC API around C++ APIs.

Reply:
> Objc files are guarded on Objc availability
What's the exact guarding mechanism used in this case?

Reply:
Maybe `#if defined(__OBJC__)` or some custom defined macro.

Reply:
Would those be `.m` or `.c` files having such macro? In case those are `.m` files, they would be detected by Clang as Objective-C sources anyway, so I think those should be excluded by some other means.

Reply:
I was thinking in `.m` and maybe even in headers where ObjC syntax is used. 

> so I think those should be excluded by some other means.

What are some of the other ways?

I realize this is similar to mixing in C++ but keeping interop mode disabled. In that case, `#ifdef __cplusplus` blocks should be used to guard C++ code within the mixed target ([forums post](https://forums.swift.org/t/se-0403-package-manager-mixed-language-target-support/66202/4)).

Reply:
IMO, the compiler should probably just produce better diagnostics here.

Reply:
Do we have an issue to track diagnostics improvement for this in the compiler?

Reply:
Here is an existing Swift issue that's tracking better diagnostics wrt ObjC interop on Linux: https://github.com/apple/swift/issues/50879. I think this covers it from the Swift side of things.

For compiling ObjC on Linux, I browsed the Clang compiler issues, but didn't see any that jumped out. I'm seeing things that indicate there are ways to get Foundation headers building on Linux ([stack overflow](https://stackoverflow.com/a/22246993/9331576)) so I'm not sure what is the preferred, if any, diagnostic here.

Reply:
My understanding is that using the Swift toolchain, you do not get an ObjC runtime on e.g. Linux. I think there needs to be a source of truth of that information somewhere external to SwiftPM that we can query to convey that to the user.

This does seem pretty orthogonal to the PR at hand, though, since getting the "Foundation.h" related diagnostic is probably the status quo for C-language family targets today, right?

Reply:
> This does seem pretty orthogonal to the PR at hand, though, since getting the "Foundation.h" related diagnostic is probably the status quo for C-language family targets today, right?

If I'm understanding this correctly: I think it's the status quo for only Objective-C/Objective-C++ files since I don't believe a Foundation header can be imported into a C-language family source file that isn't Objective-C/Objective-C++.